### ISSUE TITLE
Compilation Error - Can't find OpenNI.h

### ISSUE THREAD

**Issue by thiw1ka** [MisoRobotics/xtion2_ros#1]
Hi,

I have been trying to use your Ros package for my xtion2 but when I try to compile it in my catkin workspace, it gives me the error(error 1) saying can't find OpenNI.h. So i realize it cannot find the library directory therefore I changed cmake file's following line with {PC_OPENNI2_INCLUDE_DIRS}. Then the package got complied but failed because of some undefined references. (error 2)

I am using ros melodic with ubuntu 18.04 and I have openNI2 libraries installed. 

Appreciate if you can help me to resolve this issue.

Thanks 

Error 1 
------

In file included from /home/{your_name}/catkin_ws/src/xtion2_ros-master/src/io_interface.cxx:4:0:
/home/{your_name}/catkin_ws/src/xtion2_ros-master/include/xtion2_ros/io_interface.h:12:10: fatal error: OpenNI.h: No such file or directory
 #include <OpenNI.h>
          ^~~~~~~~~~
compilation terminated.
xtion2_ros-master/CMakeFiles/xtion2_ros.dir/build.make:86: recipe for target 'xtion2_ros-master/CMakeFiles/xtion2_ros.dir/src/io_interface.cxx.o' failed
make[2]: *** [xtion2_ros-master/CMakeFiles/xtion2_ros.dir/src/io_interface.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/{your_name}/catkin_ws/src/xtion2_ros-master/src/camera_publisher.cxx:8:0:
/home/{your_name}/catkin_ws/src/xtion2_ros-master/include/xtion2_ros/io_interface.h:12:10: fatal error: OpenNI.h: No such file or directory
 #include <OpenNI.h>
          ^~~~~~~~~~
compilation terminated.
xtion2_ros-master/CMakeFiles/xtion2_ros.dir/build.make:62: recipe for target 'xtion2_ros-master/CMakeFiles/xtion2_ros.dir/src/camera_publisher.cxx.o' failed
make[2]: *** [xtion2_ros-master/CMakeFiles/xtion2_ros.dir/src/camera_publisher.cxx.o] Error 1
CMakeFiles/Makefile2:3803: recipe for target 'xtion2_ros-master/CMakeFiles/xtion2_ros.dir/all' failed
make[1]: *** [xtion2_ros-master/CMakeFiles/xtion2_ros.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j16 -l16" failed

Error 2
---------

[100%] Linking CXX executable /home/{your_name}/catkin_ws/devel/lib/xtion2_ros/xtion2_ros_node
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::VideoFrameRef::release()':
/usr/include/openni2/OpenNI.h:591: undefined reference to `oniFrameRelease'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::VideoStream::start()':
/usr/include/openni2/OpenNI.h:786: undefined reference to `oniStreamStart'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::VideoStream::isPropertySupported(int) const':
/usr/include/openni2/OpenNI.h:1150: undefined reference to `oniStreamIsPropertySupported'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::VideoStream::_setHandle(_OniStream*)':
/usr/include/openni2/OpenNI.h:1212: undefined reference to `oniStreamGetSensorInfo'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::Device::isPropertySupported(int) const':
/usr/include/openni2/OpenNI.h:1579: undefined reference to `oniDeviceIsPropertySupported'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::Device::isCommandSupported(int) const':
/usr/include/openni2/OpenNI.h:1618: undefined reference to `oniDeviceIsCommandSupported'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::OpenNI::initialize()':
/usr/include/openni2/OpenNI.h:2099: undefined reference to `oniInitialize'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::OpenNI::shutdown()':
/usr/include/openni2/OpenNI.h:2108: undefined reference to `oniShutdown'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::OpenNI::getExtendedError()':
/usr/include/openni2/OpenNI.h:2134: undefined reference to `oniGetExtendedError'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::VideoStream::create(openni::Device const&, openni::SensorType)':
/usr/include/openni2/OpenNI.h:2620: undefined reference to `oniDeviceCreateStream'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::VideoStream::destroy()':
/usr/include/openni2/OpenNI.h:2653: undefined reference to `oniStreamDestroy'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::Device::open(char const*)':
/usr/include/openni2/OpenNI.h:2671: undefined reference to `oniDeviceOpen'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::Device::_setHandle(_OniDevice*)':
/usr/include/openni2/OpenNI.h:2714: undefined reference to `oniDeviceGetInfo'
CMakeFiles/xtion2_ros_node.dir/src/xtion2_ros_node.cxx.o: In function `openni::Device::close()':
/usr/include/openni2/OpenNI.h:2740: undefined reference to `oniDeviceClose'
/home/{your_name}/catkin_ws/devel/lib/libxtion2_ros.so: undefined reference to `oniStreamGetProperty'
/home/{your_name}/catkin_ws/devel/lib/libxtion2_ros.so: undefined reference to `oniStreamReadFrame'
/home/{your_name}/catkin_ws/devel/lib/libxtion2_ros.so: undefined reference to `oniWaitForAnyStream'
collect2: error: ld returned 1 exit status
xtion2_ros-master/CMakeFiles/xtion2_ros_node.dir/build.make:171: recipe for target '/home/{your_name}/catkin_ws/devel/lib/xtion2_ros/xtion2_ros_node' failed
make[2]: *** [/home/{your_name}/catkin_ws/devel/lib/xtion2_ros/xtion2_ros_node] Error 1
CMakeFiles/Makefile2:3865: recipe for target 'xtion2_ros-master/CMakeFiles/xtion2_ros_node.dir/all' failed
make[1]: *** [xtion2_ros-master/CMakeFiles/xtion2_ros_node.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j16 -l16" failed

**Comment by rsinnet** [comment]
I'm guessing you set the include path but didn't specify the linking path as well?
The `include_directories` command allows you to set the include path (`-I` flag passed to compiler). The `link_directories` command allows you to set the linker path (`-L` flag passed to linker).

Could you paste the diff of the changes you've made? (`git diff origin/master`)

**Comment by thiw1ka** [comment]
Hi I was able to compile the package successfully. As you mentioned I have done same change to the linking library. Please see my cmake below. I am sorry I am unable to do git diff at the moment.

However when I launch the package it's only able to open one camera. any idea where I did wrong?

-----------------------------------------------------------------------------------------------------------
 cmake_minimum_required(VERSION 2.8.3)
project(xtion2_ros)

set(CMAKE_BUILD_TYPE Debug)

add_compile_options(-std=c++14)

find_package(catkin REQUIRED
  COMPONENTS
  camera_info_manager
  cmake_modules
  cv_bridge
  image_transport
  roscpp
  roslint
)

set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_LIST_DIR}/cmake)
find_package(Boost REQUIRED COMPONENTS system)
find_package(OpenCV 3.2 REQUIRED)
find_package(OpenNI2 REQUIRED)

catkin_package(
  INCLUDE_DIRS include
  LIBRARIES ${PROJECT_NAME}
  CATKIN_DEPENDS camera_info_manager cv_bridge image_transport roscpp
  DEPENDS Boost OpenCV OpenNI2
)

roslint_cpp()

include_directories(
  include
  ${catkin_INCLUDE_DIRS}
  ${OpenCV_INCLUDE_DIRS}
  ${PC_OPENNI2_INCLUDE_DIRS}
)


set(LINK_LIBS  ${PC_OPENNI2_LIBRARIES} ${OpenCV_LIBRARIES} ${catkin_LIBRARIES})
set(NODE ${PROJECT_NAME}_node)

add_library(${PROJECT_NAME}
  src/camera_publisher.cxx
  src/io_interface.cxx
)
add_dependencies(${PROJECT_NAME} ${catkin_EXPORTED_TARGETS})
target_link_libraries(${PROJECT_NAME} ${LINK_LIBS})

add_executable(${NODE} src/${NODE}.cxx)
add_dependencies(${NODE} ${catkin_EXPORTED_TARGETS})
target_link_libraries(${NODE} ${PROJECT_NAME} ${LINK_LIBS})

------------------------------------------------------------------------------------------------------------------------
Error 
-----------------------------------------------------------------------------------------------------------------------

... logging to /home/thiw/.ros/log/2f331410-ec11-11e8-b9d5-309c231f9233/roslaunch-bender-10405.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://bender:34379/

SUMMARY
========

PARAMETERS
 * /rosdistro: melodic
 * /rosversion: 1.14.3
 * /xtion2_ros_node/device_uri: 

NODES
  /xtion2/depth/
    image_proc (image_proc/image_proc)
  /xtion2/rgb/
    image_proc (image_proc/image_proc)
  /
    xtion2_ros_node (xtion2_ros/xtion2_ros_node)

ROS_MASTER_URI=http://localhost:11311

process[xtion2_ros_node-1]: started with pid [10420]
process[xtion2/rgb/image_proc-2]: started with pid [10421]
process[xtion2/depth/image_proc-3]: started with pid [10437]
Video Camera at /dev/video0 Detected
Cannot identify '/dev/video1': 2, No such file or directory
Can't open depth camera!
[xtion2_ros_node-1] process has died [pid 10420, exit code 1, cmd /home/thiw/catkin_ws/devel/lib/xtion2_ros/xtion2_ros_node __name:=xtion2_ros_node __log:=/home/thiw/.ros/log/2f331410-ec11-11e8-b9d5-309c231f9233/xtion2_ros_node-1.log].
log file: /home/thiw/.ros/log/2f331410-ec11-11e8-b9d5-309c231f9233/xtion2_ros_node-1*.log

**Comment by rsinnet** [comment]
Are you plugging the camera into USB 3? It will _not_ work on USB 3. This is a known issue with the camera itself. You have to plug into a USB 2 port.

In terms of the above `CMakeLists.txt`, I see you're using the `target_link_libraries` command which adds `-l` flag for each library, thereby telling the linker to use this library. However, you should also do use the `link_directories` command to let the linker know where to look for these libraries: https://cmake.org/cmake/help/v3.0/command/link_directories.html

If it's in `/usr/lib` or something standard, you may not need it which it seems may be the case. It looks like your linker path is now set up correctly though since you've built successfully.

**Comment by thiw1ka** [comment]
Yes I was using USB 3 port. However when I connect it to USB 2 it will open the second camera not the first one. following is the error message. 


... logging to /home/thiw/.ros/log/fe177062-ec3b-11e8-b9d5-309c231f9233/roslaunch-bender-14174.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://bender:37505/

SUMMARY
========

PARAMETERS
 * /rosdistro: melodic
 * /rosversion: 1.14.3
 * /xtion2_ros_node/device_uri: 

NODES
  /xtion2/depth/
    image_proc (image_proc/image_proc)
  /xtion2/rgb/
    image_proc (image_proc/image_proc)
  /
    xtion2_ros_node (xtion2_ros/xtion2_ros_node)

auto-starting new master
process[master]: started with pid [14185]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to fe177062-ec3b-11e8-b9d5-309c231f9233
process[rosout-1]: started with pid [14196]
started core service [/rosout]
process[xtion2_ros_node-2]: started with pid [14200]
process[xtion2/rgb/image_proc-3]: started with pid [14204]
process[xtion2/depth/image_proc-4]: started with pid [14205]
Video Camera at /dev/video1 Detected
Cannot identify '/dev/video0': 2, No such file or directory
Can't open depth camera!
[xtion2_ros_node-2] process has died [pid 14200, exit code 1, cmd /home/thiw/catkin_ws/devel/lib/xtion2_ros/xtion2_ros_node __name:=xtion2_ros_node __log:=/home/thiw/.ros/log/fe177062-ec3b-11e8-b9d5-309c231f9233/xtion2_ros_node-2.log].
log file: /home/thiw/.ros/log/fe177062-ec3b-11e8-b9d5-309c231f9233/xtion2_ros_node-2*.log

==============================================================\
Actually initially it didn't recognized the cameras then I found online that  libSenDuck.so from asus driver files need to be copied to usr/openni2/drivers. Then only it able to detect one camera. Have you copied the same to openni2?

**Comment by rsinnet** [comment]
Yes, I remember seeing that step and I believe it was necessary although my recollection isn't perfect. I believe I roughly followed these instructions: https://answers.ros.org/question/269961/openni2_launch-asus-xtion-2-no-devices-connected/?answer=270098#post-id-270098

Also, this has only been tested on Xenial with Kinetic, but I suspect this is not an issue here. From the previous message, it seems like a USB initialization step issue. You might try unplugging and plugging back in a few times as well. After you plug in, try doing `ls /dev/video*` and see how many cameras appear. I can't remember, but you might be able to view the streams in Cheese or VLC.

You could also try using `udevadm` to interrogate the USB device and see if anything seems weird: https://linux.die.net/man/8/udevadm

**Comment by rsinnet** [comment]
@thiw1ka, were you able to resolve this?

**Comment by thiw1ka** [comment]
> @thiw1ka, were you able to resolve this?

Nope am kind of stuck here. As you mentioned it worked in the VLC but not in the cheese.

**Comment by rsinnet** [comment]
So you're able to see all the camera streams in VLC? If that's the case, you're probably close to getting it working.

**Comment by thiw1ka** [comment]
Actually the RGB camera works on VLC. But other camera doesn't show any pictures, just some vertical lines. I think its with the drivers. And am kinda confuse of udevadm. what sort of details I can get from it.

Also can you recall any changes you have done to libraries or openni2, others than you have mentioned earlier?
